internal/compare/compare.go

```diff
@@ -0,0 +1,140 @@
+	File    string // File where error was discovered
+	Message string // Message contains error message.
+}
+
+func (d *Diagnostic) String() string {
+	return fmt.Sprintf("%s:%s", d.File, d.Message)
+}
+
+// Pass provides all reported errors.
+type Pass struct {
+	lints []Diagnostic
+}
+
+// Report reports an error.
+func (p *Pass) Report(d Diagnostic) {
+	p.lints = append(p.lints, d)
+}
+
+// Lints returns all errors.
+func (p *Pass) Lints() []Diagnostic {
+	return p.lints
+}
+
+func (p *Pass) String() string {
+	var b strings.Builder
+	for _, l := range p.lints {
+		_, _ = fmt.Fprintf(&b, "%s\n", l.String())
+	}
+
+	return b.String()
+}
+
+// CompareModules looks for removed methods and added required fields.
+func (p *Pass) CompareModules(from, to *compile.Module) {
+	for name, fromService := range from.Services {
+		p.service(fromService, to.Services[name])
+	}
+
+	for n, fromType := range from.Types {
+		p.typ(fromType, to.Types[n], filepath.Base(from.ThriftPath))
+	}
+}
+
+func (p *Pass) typ(from, to compile.TypeSpec, file string) {
+	if f, ok := from.(*compile.StructSpec); ok {
+		t, ok := to.(*compile.StructSpec)
+		if !ok {
+			// This is a new Type, which is backwards compatible.
+			return
+		}
+		p.structSpecs(f, t, file)
+	}
+}
+
+// StructSpecs compares two structs defined in a Thrift file.
+func (p *Pass) structSpecs(from, to *compile.StructSpec, file string) {
+	fields := make(map[int16]*compile.FieldSpec, len(from.Fields))
+	// Assume that these two should be compared.
+	for _, f := range from.Fields {
+		// Capture state of all fields here.
+		fields[f.ID] = f
+	}
+
+	for _, toField := range to.Fields {
+		if fromField, ok := fields[toField.ID]; ok {
+			fromRequired := fromField.Required
+			toRequired := toField.Required
+			if !fromRequired && toRequired {
+				p.Report(Diagnostic{
+					File: file,
+					Message: fmt.Sprintf(
+						"changing an optional field %s in %s to required is not backwards compatible",
+						toField.ThriftName(), to.ThriftName()),
+				})
+			}
+		} else if toField.Required {
+			p.Report(Diagnostic{
+				File: file,
+				Message: fmt.Sprintf("adding a required field %s to %s is not backwards compatible",
+					toField.ThriftName(), to.ThriftName()),
+			})
+		}
+	}
+}
+
+func (p *Pass) service(from, to *compile.ServiceSpec) {
+	if to == nil {
+		// Service was deleted, which is not backwards compatible.
+		p.Report(Diagnostic{
+			File:    filepath.Base(from.File), // toModule could have been deleted.
+			Message: fmt.Sprintf("deleting service %s is not backwards compatible", from.Name),
+		})
+
+		return
+	}
+	for n := range from.Functions {
+		p.function(to.Functions[n], n, filepath.Base(from.File), from.Name)
+	}
+}
+
+func (p *Pass) function(to *compile.FunctionSpec, fn string, file string, service string) {
+	if to == nil {
+		p.Report(Diagnostic{
+			File:    file,
+			Message: fmt.Sprintf("removing method %s in service %s is not backwards compatible", fn, service),
+		})
+	}
+}
```

Comment:
In case you missed https://github.com/thriftrw/thriftrw-go/pull/538#discussion_r714340252,
we'll want to check the function parameters with the same checks as struct fields.

Deferral to a follow up diff okay, but I want to make sure you saw this.